CONTRIBUTING.md

```diff
@@ -1,61 +1,259 @@
+    ```
+
+#### Step 3: Making changes to Chapter codebase :fire:
+
+> **Note: Always follow the below steps before you start coding or working on an issue.**
+
+You are now almost ready to make changes to files but before that you should **always** follow these steps:
+
+1. Validate that you are on the `master` branch
+
+    ```sh
+    git status
+    ```
 
-## We are just getting started
+    You should get an output like this:
 
-Implementation details are still being discussed and the project is currently at a very early stage.  The stage of the project is changing on a daily basis, thoughts and ideas are being contributed at a very high pace.
+    ```sh
+    On branch master
+    Your branch is up-to-date with 'origin/master'.
 
-## What do we need help with right now?
+    nothing to commit, working directory clean
+    ```
 
-We are in the very early stages of development on this new application. We value your insight and expertise.  In order to prevent duplicate issues, please search through our existing issues to see if there is one for which you would like to provide feedback. We are currently trying to consolidate many of the issues based on topics like documentation, user interface, API endpoints, and architecture.
+    If you are not on master or your working directory is not clean, resolve any outstanding files/commits and checkout `master`:
 
-## How to submit an issue
+    ```sh
+    git checkout master
+    ```
 
-- Search through current issues to verify that you are not creating a duplicate issue.
-- Use the issue template as a guide to tell us more about your issue:
-  - **Is your feature request related to a problem? Please describe.**
-      A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
-  - **Describe the solution you'd like**
-      A clear and concise description of what you want to happen.
+2. Sync the latest changes from the chapter upstream `master` branch to your local master branch. This is very important to avoid conflicts later.
 
-  - **Describe alternatives you've considered**
-      A clear and concise description of any alternative solutions or features you've considered.
+    **Note:** If you have any outstanding Pull Request that you made from the `master` branch of your fork, you will lose them at the end of this step. You should ensure your pull request is merged by a moderator before performing this step. To avoid this scenario, you should *always* work on a branch separate from master.
 
-  - **Additional context**
-      Add any other context or screenshots about the feature request here.
+    This step **will sync the latest changes** from the main repository of chapter. 
 
-- Be as descriptive as possible.
+    Update your local copy of the freeCodeCamp upstream repository:
+    ```sh
+    git fetch upstream
+    ```
 
-## Working on a change
+    Hard reset your master branch with the chapter master:
+    ```sh
+    git reset --hard upstream/master
```

Comment:
shouldn't be we should merge like `git merge upstream/master` instead of doing a hard reset as after doing a reset we have to force push. IMO, the newcomer may mess up with their branch while doing a force push given an unexpected scenario

Reply:
I'm more familiar with the fork and
[git add remote upstream https://...](https://help.github.com/en/articles/configuring-a-remote-for-a-fork)
git merge upstream/master
pattern, so shall we go with that over the git reset approach?

Reply:
```
git pull upstream master
git push origin master
```
is how I do it

Reply:
@allella  we want the master branch to be exactly identical to the upstream. Doing `git fetch` then `git merge upstream/master` actually keeps the local commits you have made to `master`. We don't want that. We want our users to work on branches and then push their changes. In this way it'll help maintainers in keeping a easy workflow for merges and resolving conflicts.

@ScottBrenner `git pull upstream master` is equivalent to `git fetch` plus `git merge upstream/master`.

Hard resetting and force pushing makes sure that your history and commits are in sync with the upstream. All the local commits made to `master` are lost. In the long run it is useful than the usual approach as it enforces contributors to work on branches as stated earlier.

Reply:
Does that run the risk of someone committing to a their local master and then running a hard reset and blowing away their work? The contrib guide is likely to be used more by newbies than veterans, so should we error on the side of the safer way?

Reply:
The main FCC repo [CONTRIBUTING.md](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md) doesn't seem to get into how to commit, so do we even want to get into describing how to do an upstream, or link to resources that describe the approach(s) we're talking about